File: src/lib/manifest/extractors/abilities.ts
import type { AbilityRecord, AbilityKind } from "../types/records";
import type { Extractor, RawTable, RawTableName } from "../types/services";
import type { RawInventoryItem } from "./rawTypes";
import {
  deriveAbilityVerbs,
  deriveSubclassAffinities,
} from "./abilityEnrichment";
import { iterItems, projectBase, toClassName, deriveElement } from "./common";

const ABILITY_CAT_RE = /\.(supers|grenades|melee|class_abilities|movement)$/;

const KIND_MAP: Record<string, AbilityKind> = {
  supers: "super",
  grenades: "grenade",
  melee: "melee",
  class_abilities: "classAbility",
  movement: "movement",
};

type LoadTable = (table: RawTableName) => Promise<RawTable>;

function isAbilityItem(item: RawInventoryItem): boolean {
  const cat = item.plug?.plugCategoryIdentifier ?? "";
  return ABILITY_CAT_RE.test(cat);
}

function resolveKind(categoryId: string): AbilityKind | null {
  const match = ABILITY_CAT_RE.exec(categoryId);
  if (!match) return null;
  return KIND_MAP[match[1]] ?? null;
}

async function extractAbilities(loadTable: LoadTable): Promise<AbilityRecord[]> {
  const itemTable = await loadTable("DestinyInventoryItemDefinition");
  const result: AbilityRecord[] = [];

  for (const item of iterItems(itemTable)) {
    if (!isAbilityItem(item)) continue;

    const cat = item.plug?.plugCategoryIdentifier ?? "";
    const kind = resolveKind(cat);
    if (!kind) continue;

    const typeName = item.itemTypeDisplayName ?? "";
    const element = deriveElement(typeName, cat);
    const classType = toClassName(item.classType);
    const description = item.displayProperties.description;

    result.push({
      ...projectBase(item),
      description,
      kind,
      classType,
      element,
      // Rebuild abilities entity cache after extractor changes (013 enrichment fields).
      subclassAffinities: deriveSubclassAffinities({
        hash: item.hash,
        plugCategoryIdentifier: cat,
        classType,
        element,
      }),
      verbs: deriveAbilityVerbs({ hash: item.hash, description }),
    });
  }

  return result;
}

export const abilitiesExtractor: Extractor<"abilities"> = {
  store: "abilities",
  extract: (loadTable) => extractAbilities(loadTable),
};

File: flutter/packages/manifest/lib/src/extractors/abilities.dart
import '../raw/raw_types.dart';
import '../types/records.dart';
import '../types/services.dart';
import '../types/stores.dart';
import 'common.dart';

final _abilityCatRe =
    RegExp(r'\.(supers|grenades|melee|class_abilities|movement)$');

class AbilitiesExtractor implements EntityExtractor {
  @override
  MvpStoreName get store => MvpStoreName.abilities;

  @override
  Future<List<Object>> extract(LoadRawTable loadTable) async {
    final itemTable = await loadTable('DestinyInventoryItemDefinition');
    final result = <AbilityRecord>[];

    for (final item in iterItems(itemTable)) {
      if (!_isAbilityItem(item)) continue;
      final cat = item.plug?['plugCategoryIdentifier'] as String? ?? '';
      final kind = _resolveKind(cat);
      if (kind == null) continue;
      final typeName = item.itemTypeDisplayName ?? '';
      final element = deriveElement(typeName, cat);
      final classType = toClassName(item.classType);
      final base = projectBase(item);
      result.add(
        AbilityRecord(
          hash: base.hash,
          name: base.name,
          searchName: base.searchName,
          icon: base.icon,
          description: item.displayProperties.description,
          kind: kind,
          classType: classType,
          element: element,
          // Full sandbox enrichment is deferred; empty lists keep hard path clean.
          subclassAffinities: const [],
          verbs: const [],
        ),
      );
    }
    return result;
  }
}

bool _isAbilityItem(RawInventoryItem item) {
  final cat = item.plug?['plugCategoryIdentifier'] as String? ?? '';
  return _abilityCatRe.hasMatch(cat);
}

AbilityKind? _resolveKind(String categoryId) {
  final match = _abilityCatRe.firstMatch(categoryId);
  if (match == null) return null;
  switch (match.group(1)) {
    case 'supers':
      return AbilityKind.superAbility;
    case 'grenades':
      return AbilityKind.grenade;
    case 'melee':
      return AbilityKind.melee;
    case 'class_abilities':
      return AbilityKind.classAbility;
    case 'movement':
      return AbilityKind.movement;
    default:
      return null;
  }
}
